### PR TITLE
`<DatagridBody>` creates the `<RecordContext>` instead of `<DatagridRow>`

### PR DESCRIPTION
**Pull request by adguernier**
This PR attempts to delegate the creation of a  `<RecordContext>` to `<DatagridBody>` instead of `<DatagridRow>`

### CODE REVIEW

**Review by fzaninotto — CHANGES_REQUESTED** [review]
You also need to remove the reconrdContextProvider from the DatagridRow

**Comment by adguernier** [comment]
> You also need to remove the reconrdContextProvider from the DatagridRow

Done, I pushed in a hurry

File: packages/ra-ui-materialui/src/list/datagrid/DatagridBody.tsx [hunk 1]
```diff
@@ -34,31 +34,33 @@ const DatagridBody: FC<DatagridBodyProps> = React.forwardRef(
...
+                        {
+                            className: clsx(DatagridClasses.row, {
+                                [DatagridClasses.rowEven]: rowIndex % 2 === 0,
+                                [DatagridClasses.rowOdd]: rowIndex % 2 !== 0,
+                            }),
+                            expand,
+                            hasBulkActions: hasBulkActions && !!selectedIds,
+                            hover,
+                            id: record.id ?? `row${rowIndex}`,
+                            key: record.id ?? `row${rowIndex}`,
```
Comment (fzaninotto) [inline packages/ra-ui-materialui/src/list/datagrid/DatagridBody.tsx:50]: we no longer need the `key` here

File: packages/ra-ui-materialui/src/list/datagrid/DatagridBody.tsx [hunk 2]
```diff
@@ -34,31 +34,33 @@ const DatagridBody: FC<DatagridBodyProps> = React.forwardRef(
...
-                        selectable: !isRowSelectable || isRowSelectable(record),
-                        selected: selectedIds?.includes(record.id),
-                        sx: rowSx?.(record, rowIndex),
-                        style: rowStyle?.(record, rowIndex),
-                    },
-                    children
-                )
-            )}
+            {data.map((record, rowIndex) => (
+                <RecordContextProvider value={record} key={record.id}>
```
Comment (fzaninotto) [inline packages/ra-ui-materialui/src/list/datagrid/DatagridBody.tsx:38]: Use the same `key` value as previously, i.e. `record.id ?? row${rowIndex}`

File: packages/ra-ui-materialui/src/list/datagrid/DatagridBody.tsx [hunk 3]
```diff
@@ -34,31 +34,33 @@ const DatagridBody: FC<DatagridBodyProps> = React.forwardRef(
...
-                            [DatagridClasses.rowEven]: rowIndex % 2 === 0,
-                            [DatagridClasses.rowOdd]: rowIndex % 2 !== 0,
-                        }),
-                        expand,
-                        hasBulkActions: hasBulkActions && !!selectedIds,
-                        hover,
-                        id: record.id ?? `row${rowIndex}`,
-                        key: record.id ?? `row${rowIndex}`,
-                        onToggleItem,
-                        record,
```
Comment (fzaninotto) [inline packages/ra-ui-materialui/src/list/datagrid/DatagridBody.tsx:51]: This is a breaking change: a custom DatagridRow will no longer receive the record. You must document it in the upgrade guide.

File: docs/Upgrade.md [hunk 4]
```diff
@@ -1206,8 +1206,9 @@ The deprecated `<ThemeProvider theme>` prop was removed. Use the `ThemesContext.
 
 The `data-generator-retail` package has been updated to provide types for all its records. In the process, we renamed the `commands` resource to `orders`. Accordingly, the `nb_commands` property of the `customers` resource has been renamed to `nb_orders` and the `command_id` property of the `invoices` and `reviews` resources has been renamed to `order_id`.
 
+## `<DatagridBody>` No Longer Provides `record` Prop To `<DatagridRow>`
 
-
+The `<DatagridBody>` component no longer provides a `record` prop to its `<DatagridRow>` children. Instead, it provides a `recordContext` for each row. See the [`<Datagrid body/>`](./Datagrid.md#body) documentation to learn how to create your own row component.
```
Comment (fzaninotto) [inline docs/Upgrade.md:1211]: Pleae add a diff to show how the code must be changed (before: record injection, after: useRecordContext)

File: docs/Upgrade.md [hunk 5]
```diff
@@ -1206,8 +1206,9 @@ The deprecated `<ThemeProvider theme>` prop was removed. Use the `ThemesContext.
 
 The `data-generator-retail` package has been updated to provide types for all its records. In the process, we renamed the `commands` resource to `orders`. Accordingly, the `nb_commands` property of the `customers` resource has been renamed to `nb_orders` and the `command_id` property of the `invoices` and `reviews` resources has been renamed to `order_id`.
 
+## `<DatagridBody>` No Longer Provides `record` Prop To `<DatagridRow>`
```
Comment (fzaninotto) [inline docs/Upgrade.md:1209]: Move that to the "List Components" section and make it a 3rd level

**Review by fzaninotto — CHANGES_REQUESTED** [review]

**Review by erwanMarmelab — APPROVED** [review]

File: docs/Upgrade.md [hunk 6]
```diff
@@ -759,6 +759,60 @@ Besides, the buttons passed as `bulkActionButtons` no longer receive any prop. I
...
+-                           key: record.id ?? `row${rowIndex}`,
+                            onToggleItem,
+-                           record,
+                            resource,
+                            //...
+                        },
+                        children
+-                   )
+                    )}
++               </RecordContextProvider>
```
Comment (fzaninotto) [inline docs/Upgrade.md:791]: wrong diff: you should show how a custom DatagridRow no longer reads the record from its props, but from the Recordcontext.

File: docs/Upgrade.md [hunk 7]
```diff
@@ -759,6 +759,60 @@ Besides, the buttons passed as `bulkActionButtons` no longer receive any prop. I
...
++            ))}
+        </TableBody>
+    )
+);
+```
+
+Also, `<DatagridRow>` no longer provides a `<RecordContextProvider>` to its children:
```
Comment (fzaninotto) [inline docs/Upgrade.md:798]: that's wrong

**Review by fzaninotto — CHANGES_REQUESTED** [review]

**Review by erwanMarmelab — APPROVED** [review]

File: docs/Upgrade.md [hunk 8]
```diff
@@ -759,6 +759,51 @@ Besides, the buttons passed as `bulkActionButtons` no longer receive any prop. I
 
 The deprecated `<PaginationLimit>` component was removed.
 
+### `<DatagridBody>` No Longer Provides `record` Prop To `<DatagridRow>`
+
+The `<DatagridBody>` component no longer provides a `record` prop to its `<DatagridRow>` children. Instead, it provides a `recordContext` for each row:
```
Comment (djhi) [inline docs/Upgrade.md:764]:
```suggestion
The `<DatagridBody>` component no longer provides a `record` prop to its `<DatagridRow>` children. Instead, it provides a `RecordContext` for each row: 
```

**Review by djhi — APPROVED** [review]

**Pull request merged**